File: socketd/src/main/java/org/noear/socketd/protocol/impl/ProcessorDefault.java
package org.noear.socketd.protocol.impl;

import org.noear.socketd.protocol.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;

/**
 * 处理器
 *
 * @author noear
 * @since 2.0
 */
public class ProcessorDefault implements Processor {
    private static Logger log = LoggerFactory.getLogger(ProcessorDefault.class);

    private Listener listener = new SimpleListener();


    @Override
    public void setListener(Listener listener) {
        if (listener != null) {
            this.listener = listener;
        }
    }

    public void onReceive(Channel channel, Frame frame) throws IOException {
        if (log.isTraceEnabled()) {
            log.trace("{}", frame);
        }

        if (frame.getFlag() == Flag.Connect) {
            //if server
            Message message = frame.getMessage();
            channel.setHandshaker(new Handshaker(message));
            channel.sendConnack(message); //->Connack

            onOpen(channel.getSession());
        } else if (frame.getFlag() == Flag.Connack) {
            //if client
            Message message = frame.getMessage();
            channel.setHandshaker(new Handshaker(message));

            onOpen(channel.getSession());
        } else {
            if (channel.getHandshaker() == null) {
                channel.close();
                if (log.isWarnEnabled()) {
                    log.warn("Channel handshaker is null, sessionId={}", channel.getSession().getSessionId());
                }
                return;
            }

            channel.setLiveTime();

            try {
                switch (frame.getFlag()) {
                    case Ping: {
                        channel.sendPong();
                        break;
                    }
                    case Pong: {
                        break;
                    }
                    case Close: {
                        channel.close();
                        onClose(channel.getSession());
                        break;
                    }
                    case Message: {
                        onMessage(channel.getSession(), frame.getMessage());
                        break;
                    }
                    case Request: {
                        onMessage(channel.getSession(), frame.getMessage());
                        break;
                    }
                    case Subscribe: {
                        onMessage(channel.getSession(), frame.getMessage());
                        break;
                    }
                    case Reply:
                    case ReplyEnd:{
                        channel.retrieve(frame);
                        break;
                    }
                    default: {
                        channel.close();
                        onClose(channel.getSession());
                    }
                }
            } catch (Throwable e) {
                onError(channel.getSession(), e);
            }
        }
    }


    @Override
    public void onOpen(Session session) throws IOException{
        listener.onOpen(session);
    }

    @Override
    public void onMessage(Session session, Message message) throws IOException {
        listener.onMessage(session, message);
    }

    @Override
    public void onClose(Session session) {
        listener.onClose(session);
    }

    @Override
    public void onError(Session session, Throwable error) {
        listener.onError(session, error);
    }
}

File: python/socketd/socketd/core/ProcessorDefault.py
import asyncio
from abc import ABC
from loguru import logger

from .Channel import Channel
from .Handshake import Handshake
from .Processor import Processor
from .Costants import Flag
from .SimpleListener import SimpleListener
from .module.Entity import EntityMetas
from .module.Frame import Frame


class ProcessorDefault(Processor, ABC):

    def __init__(self):
        self.listener = SimpleListener()
        self.log = logger.opt()

    def set_listener(self, listener):
        if listener is not None:
            self.listener = listener

    async def on_receive(self, channel: Channel, frame):
        self.log.debug("on_receive {frame}", frame=frame)
        if frame.get_flag() == Flag.Connect:
            connectMessage = frame.get_message()
            channel.set_handshake(Handshake(connectMessage))
            await channel.send_connack(connectMessage)
            self.on_open(channel.get_session())
        elif frame.get_flag() == Flag.Connack:
            message = frame.get_message()
            channel.set_handshake(Handshake(message))
            self.on_open(channel.get_session())
        else:
            if channel.get_handshake() is None:
                await channel.close()
                self.log.warning("Channel handshake is None, sessionId={}", channel.get_session().get_session_id())
                return

            channel.set_live_time()

            try:
                if frame.get_flag() == Flag.Ping:
                    await channel.send_pong()
                elif frame.get_flag() == Flag.Pong:
                    pass
                elif frame.get_flag() == Flag.Close:
                    await channel.close()
                    self.on_close(channel.get_session())
                elif frame.get_flag() in [Flag.Message, Flag.Request, Flag.Subscribe]:
                    self.on_receive_do(channel, frame, False)
                elif frame.get_flag() in [Flag.Reply, Flag.ReplyEnd]:
                    self.on_receive_do(channel, frame, True)
                else:
                    await channel.close()
                    self.on_close(channel.get_session())
            except Exception as e:
                logger.warning(e)
                self.on_error(channel.get_session(), e)

    def on_receive_do(self, channel: Channel, frame: Frame, isReply):
        fragmentIdxStr = frame.get_message().get_entity().get_meta(EntityMetas.META_DATA_FRAGMENT_IDX)
        if fragmentIdxStr is not None:
            index = int(fragmentIdxStr)
            frameNew = channel.get_config().get_fragment_handler().aggrFragment(channel, index, frame)
            if frameNew is None:
                return
            else:
                frame = frameNew

        if isReply:
            channel.retrieve(frame, lambda error:self.on_error(channel, error))
        else:
            self.on_message(channel, frame.get_message())

    def on_open(self, session):
        self.listener.on_open(session)

    def on_message(self, channel: Channel, message):
        # self.listener.on_message(channel.get_session(), message)
        channel.get_config().get_executor().submit(lambda _message: asyncio.run(self.listener.on_message(channel.get_session(), _message)), message)

    def on_close(self, session):
        self.listener.on_close(session)

    def on_error(self, session, error):
        self.listener.on_error(session, error)
